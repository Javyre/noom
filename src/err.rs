use ::owo_colors::{OwoColorize, Stream::Stderr, Style};

#[derive(Debug, Clone, Copy)]
pub struct Location {
    line: u32,
    col: u32,
    offset: u32,
}

impl<'s> From<crate::par::Span<'s>> for Location {
    fn from(span: crate::par::Span<'s>) -> Self {
        Self {
            line: span.location_line(),
            col: span.get_utf8_column() as u32,
            offset: span.location_offset() as u32,
        }
    }
}

#[derive(Debug)]
pub struct Error(pub Location, pub &'static str);

impl Error {
    pub fn print(&self, fname: &str, file: &str) {
        let Location { line, col, .. } = self.0;
        let msg = self.1;

        let err_style = Style::new().red().bold();

        eprint!(
            "{fname}:{line}:{col}:{}{msg}\n",
            "error: ".if_supports_color(Stderr, |t| t.style(err_style))
        );
        underline(file, self.0, self.0);
    }
}

fn underline(file: &str, beg: Location, end: Location) {
    // TODO: implement multi-line ranges
    assert_eq!(beg.line, end.line);

    // Find beggining and end of line.
    let line_begin = file[0..beg.offset as usize]
        .char_indices()
        .into_iter()
        .rev()
        .find(|&(_, c)| c == '\n' || c == '\r')
        .map(|(i, _)| i + 1)
        .unwrap_or(0);
    let line_end = end.offset as usize
        + file[end.offset as usize..]
            .char_indices()
            .into_iter()
            .find(|&(_, c)| c == '\n' || c == '\r')
            .map(|(i, _)| i)
            .unwrap_or(file[end.offset as usize..].len());

    eprint!("| {}\n", &file[line_begin..line_end]);

    let underline_style = Style::new().green().bold();
    let underline = format!("{0:^<1$}", "", ((end.col - beg.col) as usize).max(1));

    eprint!(
        "| {0: <1$}{2}\n",
        "",
        beg.col as usize - 1,
        underline.if_supports_color(Stderr, |t| t.style(underline_style))
    );
}

pub fn print_file_errs(fname: &str, file: &str, errs: &Vec<Error>) {
    for e in errs {
        e.print(fname, file);
    }
}
