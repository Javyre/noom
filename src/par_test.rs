#[cfg(test)]
fn parse_snap<'s>(prog: &'s str) -> crate::par::Block<'s> {
    use std::io::Write;

    use crate::err;
    use crate::par;

    let (chunk, errs) = par::parse_chunk(prog);

    let mut stderr = std::io::stderr();
    err::write_file_errs(&mut stderr, "-", &prog, &errs).unwrap();
    stderr.flush().unwrap();

    assert_eq!(errs.len(), 0);

    chunk
}

macro_rules! test_par {
    ($name:ident, $prog:expr) => {
        #[test]
        fn $name() {
            insta::assert_yaml_snapshot!(parse_snap($prog));
        }
    };
}

test_par!(
    test_table,
    r"
let foo = {
    { abc = 'something', 123, 123 },
    { [2] = 'something', 1, 123 },
    { [asd] = 'other', 123, 544 }
};
"
);

test_par!(
    test_func_assign,
    r"
f(a: number): number = a * 2;
"
);

test_par!(
    test_func_let,
    r"
let f(a: number): number = a * 2;
"
);

test_par!(
    test_func_let_2,
    r"
let f = .(a: number): number { a * 2 };
"
);

test_par!(
    test_func_let_3,
    r"
let f: (number): number = .(a){ a * 2 };
"
);

test_par!(
    test_if,
    r"
if (1 + a > foo) .{
    then_()
} else .{
    else_()
};
"
);

test_par!(
    test_if_2,
    r"
if (1 + a > foo) .{
    then_()
};
"
);
