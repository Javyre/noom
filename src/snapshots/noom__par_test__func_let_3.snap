---
source: src/par_test.rs
expression: "parse_snap(r\"\nlet f: (number) -> number = fn (a){ a * 2 };\n\")"
---
stmts:
  - Let:
      - span:
          line: 2
          col: 5
          fragment: f
      - Func:
          - - TypeIdent:
                - line: 2
                  col: 9
                  fragment: number
                - []
          - TypeIdent:
              - line: 2
                col: 20
                fragment: number
              - []
      - Func:
          - - - span:
                  line: 2
                  col: 33
                  fragment: a
              - ~
          - Block:
              stmts: []
              ret:
                BinaryOp:
                  - Ident:
                      span:
                        line: 2
                        col: 37
                        fragment: a
                  - line: 2
                    col: 39
                    fragment: "*"
                  - Number:
                      span:
                        line: 2
                        col: 41
                        fragment: "2"
          - ~
ret: ~

